python/whylogs/experimental/core/metrics/embedding_metric.py

```diff
@@ -0,0 +1,216 @@
+
+
+def _serialize_ndarray(a: np.ndarray) -> bytes:
+    bio = BytesIO()
+    np.save(bio, a, allow_pickle=False)
+    return bio.getvalue()
+
+
+def _deserialize_ndarray(a: bytes) -> np.ndarray:
+    bio = BytesIO(a)
+    return np.load(bio, allow_pickle=False)
+
+
+class MatrixComponent(MetricComponent[np.ndarray]):
+    # mtype = np.ndarray
+    type_id = 101
+
+
+@serializer(type_id=101)
+def serialize(value: np.ndarray) -> MetricComponentMessage:
+    return MetricComponentMessage(serialized_bytes=_serialize_ndarray(value))
+
+
+@deserializer(type_id=101)
+def deserialize(msg: MetricComponentMessage) -> np.ndarray:
+    return _deserialize_ndarray(msg.serialized_bytes)
+
+
+class DistanceFunction(Enum):
+    euclidean = euclidean_distances
+    cosine = cosine_distances
+
+
+@dataclass(frozen=True)
+class EmbeddingConfig(MetricConfig):
+    references: np.ndarray = field(default_factory=lambda: np.zeros((1, 1)))  # rows are reference vectors
+    labels: Optional[List[str]] = None
+    distance_fn: DistanceFunction = DistanceFunction.cosine
+    serialize_references: bool = True
+
+    # TODO: limit refeence size
+
+    def __post_init__(self) -> None:
+        if len(self.references.shape) != 2:
+            raise ValueError("Embedding reference matrix must be 2 dimensional")
+
+        if self.labels:
+            if len(self.labels) != self.references.shape[0]:
+                raise ValueError(
+                    f"Number of labels ({len(self.labels)}) must match number of reference vectors ({self.references.shape[1]})"
+                )
+
+
+@dataclass
+class EmbeddingMetric(MultiMetric):
+    references: MatrixComponent
+    labels: List[str]
+    distance_fn: DistanceFunction
+    serialize_references: bool
+
+    def __post_init__(self):
+        submetrics = {
+            f"{label}_distance": {
+                "distribution": StandardMetric.distribution.zero(),
+                "counts": StandardMetric.counts.zero(),
+                "types": StandardMetric.types.zero(),
+            }
+            for label in self.labels
+        }
+        submetrics.update(
+            {
+                "closest": {
+                    "frequent_items": StandardMetric.frequent_items.zero(),
+                    "counts": StandardMetric.counts.zero(),
+                    "types": StandardMetric.types.zero(),
+                }
+            }
+        )
+        super().__init__(submetrics)
+        # sort labels and permute reference matrix rows to match to support deserialization
+        label_indices = np.argsort(self.labels).tolist()
+        self.labels = sorted(self.labels)
+        self.references = MatrixComponent(self.references.value[label_indices, :])
+
+    @property
+    def namespace(self) -> str:
+        return "embedding"
+
+    def merge(self, other: "EmbeddingMetric") -> "EmbeddingMetric":
+        if self.references.value.shape != other.references.value.shape:
+            if other.references.value.shape == (1, 1):
```

Comment:
This still doesn't work for merging with deserialized metrics that didn't include their reference matrices. I'll fix it later.

Reply:
Fair enough. Can you add a comment with TODO or file an issue?